plugins/modules/kea_command.py

```diff
@@ -0,0 +1,216 @@
+  - community.general.attributes
+  - community.general.attributes.platform
+attributes:
+  check_mode:
+    support: none
+  diff_mode:
+    support: none
+  platform:
+    support: full
+    platforms: posix
+"""
+
+EXAMPLES = r"""
+vars:
+  ipaddr: "192.168.123.45"
+  hwaddr: "00:00:5E:00:53:00"
+tasks:
+
+  # an example for a request acquiring information
+  - name: Get KEA DHCP6 status
+    kea_command:
+      command: status-get
+      rv_unchanged: [0]
+      socket: /run/kea/kea6-ctrl-socket
+    register: kea6_status
+  - name: Display registered status result
+    ansible.builtin.debug:
+      msg: KEA DHCP6 running on PID {{ kea6_status.response.arguments.pid }}
+
+  # an example for requests modifying state
+  - name: Remove existing leases for {{ ipaddr }}, if any
+    kea_command:
+      command: lease4-del
+      arguments:
+        ip-address: "{{ ipaddr }}"
+      rv_changed: [0]
+      rv_unchanged: [3]
+  - name: Add DHCP lease for {{ ipaddr }}
+    kea_command:
+      command: lease4-add
+      arguments:
+        ip-address: "{{ ipaddr }}"
+        hw-address: "{{ hwaddr }}"
+      rv_changed: [0]
+"""
+
+RETURN = r"""
+response:
+  description: The server JSON response.
+  returned: when available
+  type: dict
+"""
+
+import json
+import os
+import socket
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def _parse_constant(s):
+    raise ValueError('Invalid JSON: "%s"' % s)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            command=dict(type='str', required=True),
+            arguments=dict(type='dict'),
+            rv_unchanged=dict(type='list', elements='int', default=[]),
+            rv_changed=dict(type='list', elements='int', default=[]),
+            socket=dict(type='path', default='/run/kea/kea4-ctrl-socket'),
+        ),
+    )
+
+    cmd = {}
+    cmd['command'] = module.params['command']
+    if module.params['arguments'] is not None:
+        cmd['arguments'] = module.params['arguments']
+    cmdstr = json.dumps(cmd, ensure_ascii=True, allow_nan=False, indent=None, separators=(',', ':'), sort_keys=True)
+    rvok = module.params['rv_unchanged']
+    rvch = module.params['rv_changed']
+    sockfn = module.params['socket']
+
+    r = {
+        'changed': False
+    }
+    rsp = b''
+
+    if not os.path.exists(sockfn):
+        r['msg'] = 'socket (%s) does not exist' % sockfn
+        module.fail_json(**r)
+
+    with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
+        try:
+            sock.connect(sockfn)
+        except OSError as ex:
+            r['msg'] = 'could not connect to socket (%s)' % sockfn
+            r['exception'] = ex
+            module.fail_json(**r)
+        # better safe in case anything fails…
+        r['changed'] = True
+        try:
+            sock.sendall(cmdstr.encode('ASCII'))
+        except OSError as ex:
+            r['msg'] = 'could not write to socket (%s)' % sockfn
+            r['exception'] = ex
```

Comment:
The `exception` return value is a misnomer - it is not the `Exception` but rather the traceback of the exception.

See https://docs.ansible.com/ansible/latest/reference_appendices/common_return_values.html#exception

Also, it might be possible that the `OSError` popping up already has a good description of what happened, so rewriting `msg` might be a bit redundant. Maybe `r['msg'] = str(ex)` could do the trick? I have not tested that, but might be worth a look - if that works, instead of 3 `try/except` blocks, you could have the entire socket logic within one single `try` block.

Reply:
> The `exception` return value is a misnomer - it is not the `Exception` but rather the traceback of the exception.

Oh, so the message is lost?

In this case, we’ll definitely want the `str(ex)`… but I think we’ll also want the phase. I think I can still simplify it a bit while keeping the phase though. Will push an update.

Reply:
Well, kinda, yes. The Ansible machinery expects `exception` to be a `traceback` object, so not having the expected outcome given by the name.